scenario/state.py

```diff
@@ -643,24 +644,35 @@ def _generate_new_change_id():
 
 
 @dataclasses.dataclass(frozen=True)
-class ExecOutput(_max_posargs(0)):
+class Exec(_max_posargs(1)):
     """Mock data for simulated :meth:`ops.Container.exec` calls."""
 
+    command_prefix: Sequence[str]
     return_code: int = 0
     """The return code of the process (0 is success)."""
     stdout: str = ""
     """Any content written to stdout by the process."""
     stderr: str = ""
     """Any content written to stderr by the process."""
+    stdin: Optional[str] = None
```

Comment:
in order to avoid confusions, how about we make this private and add a `stdin` property?

Also, conceptually, I feel that `stdin` does not quite belong to the State but to Context and is more on the level of `juju_log`.

Just spitballing here, but how do we feel about:

```python
ctx = Context(MyCharm)
exe = MyExec(['ls', '-ll'])
ctr = Container("grafana", execs=[exe])
ctx.run('update-status', State(containers={ctr}))

# give me the stdin the charm sent to the grafana container while executing this command
stdin: str = ctx.get_stdin(ctr, exe)
```

Reply:
> in order to avoid confusions, how about we make this private and add a stdin property?

Yes, that's much better, thanks! Done. (Although it may end up being irrelevant anyway!)

> Also, conceptually, I feel that stdin does not quite belong to the State but to Context and is more on the level of juju_log.

I feel like it's much tidier to have everything bundled up together in the one mock process object rather than split off the input to it into the context.

In addition, it does seem reasonably likely that we'd want to extend this in the future to handle the other 'input' elements of `exec()` - to verify that the right envars are being set, or that it's being run in the right working directory, or as the right user, and so on. It would be fairly clean to add those as additional private attributes of `Exec` and properties to expose them. I don't think we would want all of them as additional context methods, or to have `ctx.get_exec_input(container, command)` return some new object (essentially `ops.testing.ExecArgs`) that collects them all.

I also don't love how the `Context` object is becoming more and more context+side_effects.

I do get that it's not in the output state in the way that changes to secrets or relation data are - once the process has finished, it's gone. It somewhat feels like stdout/stderr/return_code are the same, in that the executable presumably still lives in the container but what happens when you (mock) run it isn't really "state".

I spent a few minutes convincing the rest of charm-tech that it should stay as it is in our daily call today, and then as I've tried to write it out I've found myself disagreeing with my own arguments and trying to come up with alternatives (I will spare you the terrible ones!).

An executable is basically a method that takes some input and returns some output. What if we did something similar to the `handle_exec` system from Harness instead?

```python
class MyCharm(ops.CharmBase):
    ...
    def _on_update_status(self, event):
        container = self.unit.get_container("foo")
        grep = container.exec(["grep", "scenario"])
        grep.stdin.write("foo\nsome text the charm wrote in the scenario test\nfoo")
        out, err = grep.wait_output()
        # do something with out, err

def mock_grep(args: ops.testing.ExecArgs, ) -> ops.testing.ExecResult:
    assert args.command == ["grep", "scenario"]
    assert args.stdin == "foo\nsome text the charm wrote in the scenario test\nfoo"
    return ops.testing.ExecResult(stdout="some text the charm wrote in the scenario test")

ctx = Context(MyCharm)
grep = Exec(["grep"], mock_grep)
container = Container("foo", execs={grep})
state_in = State(containers={container})

state_out = ctx.run(ctx.on.update_status(), state_in)

assert ...
```

So the mock executable lives on in the state, but the input and output is something that's is re-run each time. The input, output, and exit code all live together in one (temporary) place.

I'm not a huge fan of having some of the asserts end up in the mock function, breaking the very clean typical pattern of arrange/act/assert. However, you could still do that with `nonlocal` or like:

```python
command = []
input = []

def mock_grep(args: ops.testing.ExecArgs) -> ops.testing.ExecResult:
    command.append(args.command)
    input.append(args.stdin)
    return ops.testing.ExecResult(stdout="some text the charm wrote in the scenario test")

ctx = Context(MyCharm)
grep = Exec(["grep"], mock_grep)
container = Container("foo", execs={grep})
state_in = State(containers={container})

state_out = ctx.run(ctx.on.update_status(), state_in)

assert command == ["grep", "scenario"]
assert input == ["..."]
```

I would personally leave out the "results" shortcut that Harness has. It could always be added later, and it's simple to do - `result=127` is just `lambda _: ExecResult(exit_code=127)` and `result="foo"` is just `lambda _: ExecResult(stdout="foo")`.

WDYT? Also @benhoyt since we were in agreement on a different result earlier today, sorry! I probably should have written up my draft response before talking to the team and then would have come to a different conclusion earlier.

Reply:
mmm I don't really like the proposal of doing the asserts in a temporary spot. It's very un-scenario-y.

perhaps:
```

with container.mock('ps', '-aux') as exec:
    ctx.run(event, state)
    assert exec.input == foo
    assert exec.output == foo
```

with the risk that if you need to mock lots of commands you end up with a confusing bunch of individual contexts?

Reply:
I do also like the flexibility that a method has, for things like returning different output depending on the arguments or the stdin. On the other hand, it can't be included in a JSON dump like the rest of the `State` can, which is not great - although arguably it's not that different from `Container.mounts` serialising to a map of locations of content, not the content itself.

I played around with contexts as well - they do seem to be a reasonably natural fit in terms of having something that only exists within a specific scope. It can indeed end up with a lot, e.g.:

```python
with (
    container.mock(['ps', 'aux'], stdout=PS) as ps_exec,
    container.mock(['sleep', '10]) as sleep_exec,
    container.mock(['/bin/false'], exit_code=1) as false_exec,
    container.mock(['grep', 'scenario']) as grep_exec,
):
    ctx.run(event, state)
    assert grep_exec.stdin == bar
```

It's worse in Python <3.10 where you don't get to wrap the contexts in `()`.

What about bundling it into the existing context manager?

```python
exec = Exec(["ls", "-lh"], stdout=LS_LH)
container = Container(execs={exec})
state = State(containers={container})
ctx = Context(MyCharm)
with ctx(event, state) as mgr:
    state_out = mgr.run()
    assert mgr.exec_history == [ExecArgs(...), ExecArgs(...)]
```

If you don't care about how the process is executed and just need to mock out the output/return code then you don't need to use the context manager. I do like it more when all the pieces of the process are together, but I also get that it's potentially confusing with some being input and some output (and it's kinda opposite to what you expect, since the input the charm writes is the output of the test, etc).

Reply:
mmm, quite like the idea of enclosing it in the Manager object, but I don't think that's going to be as easy to remember as having it directly on the exec.

Reply:
in terms of API discoverability that's a net loss I'm afraid

Reply:
Ok, so I've taken this back to something much closer to @PietroPasotti's earlier suggestion, as also discussed with the Charm-Tech team last Friday.

The `Exec` object only has the results of the process (exit code, stdout, stderr), just like in Scenario 6, and the input to the process (stdin, user, group, environment, etc) are available from the Context in a new `exec_history` attribute.

I feel like we're doing a great job of keeping the `State` only state but not a great job at keeping `Context` only context - it's more and more "context plus side effects". But this way of handling exec is consistent at the moment, so let's go with this way, and I'll open a different ticket for tidying up `Context`.